src/stream/mod.rs

```diff
@@ -471,3 +476,194 @@ impl Stream {
+    /// considered "low." Defaults to 0.
+    pub fn buffered_amount_low_threshold(&self) -> usize {
+        self.stream.buffered_amount_low.load(Ordering::SeqCst)
+    }
+
+    /// get_num_bytes_in_reassembly_queue returns the number of bytes of data currently queued to
+    /// be read (once chunk is complete).
+    pub(crate) async fn get_num_bytes_in_reassembly_queue(&self) -> usize {
+        // No lock is required as it reads the size with atomic load function.
+        let reassembly_queue = self.stream.reassembly_queue.lock().await;
+        reassembly_queue.get_num_bytes()
+    }
+}
+
+impl AsyncRead for PollStream<'_> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        if buf.remaining() == 0 {
+            return Poll::Ready(Ok(()));
+        }
+
+        let fut = match self.read_fut.as_mut() {
+            Some(fut) => fut,
+            None => {
+                // read into a temporary buffer because `buf` has an unonymous lifetime, which can be
+                // shorter than the lifetime of `read_fut`.
+                let stream = self.stream.clone();
+                let mut temp_buf = Vec::with_capacity(buf.capacity());
+                self.read_fut.get_or_insert(Box::pin(async move {
+                    let res = stream.read(temp_buf.as_mut_slice()).await;
+                    match res {
+                        Ok(n) => { 
+                            temp_buf.truncate(n);
+                            Ok(temp_buf) 
+                        },
+                        Err(e) => Err(e),
+                    }
+                }))
+            }
+        };
+
+        loop {
+            match fut.as_mut().poll(cx) {
+                Poll::Pending => return Poll::Pending,
+                // retry immediately upon empty data or incomplete chunks
+                // since there's no way to setup a waker.
+                Poll::Ready(Err(Error::ErrTryAgain)) => {}
+                Poll::Ready(Err(e)) => return Poll::Ready(Err(e.into())),
+                Poll::Ready(Ok(read_buf)) => {
+                    let len = std::cmp::min(read_buf.len(), buf.remaining());
+                    buf.put_slice(&read_buf[..len]);
+                    self.read_fut = None;
+                    return Poll::Ready(Ok(()));
+                }
+            }
+        }
+    }
+}
+
+impl AsyncWrite for PollStream<'_> {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        let fut = match self.write_fut.as_mut() {
+            Some(fut) => fut,
+            None => {
+                let stream = self.stream.clone();
+                let bytes = Bytes::copy_from_slice(buf);
+                self.write_fut.get_or_insert(Box::pin(async move {
+                    stream.write(&bytes).await
+                }))
```

Comment:
You should immediately return `Poll::Ready(Ok(bytes.len()))` here

Reply:
Why? It's not guaranteed that write succeeds at this point.  Even though I'm `await`ing write, await is wrapped inside `async move` which is a Future. Writing could result in Err, which will be returned upon calling `poll` below.

Reply:
`poll_write` only queues data for sending, but doesn't have to send it. It's `poll_flush` that is responsible for sending data.
`poll_write` should return `Pending` only if the queue is full and it's not possible to queue more.

Reply:
If someone calls `poll_write(some_data)` and it returns `Poll::Pending`, then it's as if nothing happened. The user could then call `poll_write(something_totally_different)` and not expect anything from `some_data` to actually be sent.

Reply:
ah, thanks for explaining. makes sense

Reply:
still I don't think we should return Poll::Ready(Ok(bytes.len())) because we have zero guarantee that data was buffered or written

Reply:
`If successful, then it must be guaranteed that n <= buf.len(). A return value of 0 typically means that the underlying object is no longer able to accept bytes and will likely not be able to in the future as well, or that the buffer provided is empty.`

Reply:
In this abstraction, `write_buf` *is* the buffer. When you call `self.write_buf = ...`, you're buffering the data.

Reply:
Ok. I think I get it now. Double checked w/ tokio's developers just in case. Will make changes now.